### ISSUE TITLE
FairRootManager->RegisterAny(...) doesn't find dictionary for std::vector<SomeClass> if using std::shared_ptr

### ISSUE THREAD

**Issue by KlewinS** [FairRootGroup/FairRoot#789]
I would like to register a `std::shared_ptr<std::vector<SomeClass>> container` with RegisterAny() of FairRootManager to write out my output data.

I already heard, that currently one needs to trick RegisterAny to register a std::shared_ptr by using something like `static auto tmpPtr = container.get();` and then registering the `tmpPtr`. This works in some cases for me.

But now I have a problem registering a `shared_ptr<vector<SomeClass>>`. If I use this trick and register my tmpPtr, no dictionary is found later during runtime while the corresponding branch should be created.
But, If I use a plain pointer `std::vector<SomeClass>* container`, then I can register it without any problem and the dictionary is found. The strange thing now, if I first register the plain pointer, and afterwards the shared_ptr (via `static auto tmpPtr = container.get()`), then also the dictionary for the shared_ptr is found and the output is filled without any problem. Can someone help me with that?

**Comment by dennisklein** [comment]
1. `static auto tmpPtr = container.get();` seems off to me. Do you really want a [static local variable](http://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables) here? Just pass `container.get()` as second argument to `RegisterAny()`.

2. I am not an expert on ROOT dictionaries, but the dictionary generation should be independent of how you call `RegisterAny()`. Are you sure, you didn't just see some effects of an outdated dictionary file from one build iteration to the next or such?

**Comment by KlewinS** [comment]
Hi,  thanks for the comment.


to 1:  I'm sure that I really _don't_ want to have a static local variable there, but passing just the `container.get()` does not work, no dictionary will be found.

to 2: This I don't really know. If someone can tell me, where to find the dictionaries, I will check.

**Comment by dennisklein** [comment]
> ... no dictionary will be found. 

Do you have an error message at hand for such a case?

**Comment by KlewinS** [comment]
Sorry, with that I was a bit inaccurate. Passing the `.get()` as second argument doesn't compile:

> error: invalid initialization of non-const reference of type 'std::vector<o2::TPC::ClusterHardwareContainerFixedSize<8192u> >* &' from an rvalue of type 'std::vector<o2::TPC::ClusterHardwareContainerFixedSize<8192u> >*

Maybe you were correct with your 2. point somehow the dictionary is found now, even without registering the plain pointer in advance.

**Comment by KlewinS** [comment]
Or maybe not, somehow now the dictionary is not found anymore... 

> Warning in <TTree::Bronch>: FairEventHeader cannot be split, resetting splitlevel to 0
CREATING BRANCH TPCClusterHW0
[runTPC][13:25:42.154551][FATAL] No dictionary found for std::vector<o2::TPC::ClusterHardwareContainerFixedSize<8192u>, std::allocator<o2::TPC::ClusterHardwareContainerFixedSize<8192u> > >

**Comment by dennisklein** [comment]
Ok, the `RegisterAny` interface requires an lvalue. Then I would expect that `auto tmpPtr = container.get();` works also without the `static`. The `static`ness should be orthogonal here.

> If someone can tell me, where to find the dictionaries, I will check.
The dictionary files end up in the same directory as the shared library files AFAIK. (That is at least the case for the install tree of FairRoot and AliceO2)

**Comment by KlewinS** [comment]
Ok, I tried, but apparently the `static` has to stay there, otherwise the pointer which is registered (`auto tmpPtr = container.get()`) is destroyed after the end of the function and I get a segmentation violation...

**Comment by dennisklein** [comment]
Yeah, this may be a bug/feature? in the `template<typename T> void RegisterAny(const char* name, T* &obj, Bool_t toFile)` interface. I don't understand yet, why the type of the second argument is `T*&` and not just `T*`. @sawenzel Can you give us some insight here?

**Comment by KlewinS** [comment]
but the actually problem is that the dictionary is not found, at least not always. May the reason be an outdated dictionary or not, apparently the only way to ensure that the dictionary is always found, is to register first a plain pointer and afterwards the shared_ptr via the `static auto tmpPtr = container.get();` construct.

**Comment by dennisklein** [comment]
> but the actually problem is that the dictionary is not found, at least not always

How are you testing this? Is it not found when you run the same executable multiple times in a row without rebuilding in between? Or is it not found sometimes, when you change something and rebuild and then run it? What kind of changes do you perform? Can you give a more concrete example when/how it fails or point me to some code I can run?

**Comment by sawenzel** [comment]
The currently supported way to write a `shared_ptr<T>` is, like described, to get the raw pointer and put into a static lvalue which is passed to ROOT.

If this doesn't work please send a log of the actual compile/runtime problem (together with the code). Are you sure you created a dictionary for `T`?

P.S. The alternative might be to explicitly create a dictionary for `shared_ptr<T>` but this I haven't tried yet.

**Comment by dennisklein** [comment]
@sawenzel What is the rational of passing a reference to a pointer instead of passing a pointer by value to `RegisterAny`?

**Comment by KlewinS** [comment]
@dennisklein no, if the dictionary is found, then it seems to be always found. Only after I recompile due to some changes It might not be found. The changes were never in the class `ClusterHardwareContainerFixedSize` (or classes it is based on), only in other classes were I use it.

I register the vectors here:
https://github.com/KlewinS/AliceO2/blob/163e30db6c28017b408a27b6a5fc4dd0a6405deb/Detectors/TPC/reconstruction/src/ClustererTask.cxx#L90

you can run this stuff for example with `runTPC -m clus` (after it was the simulation and digitisation was executed, with `-m sim` and `-m digi`) after compiling O2

Sometimes I see the error after commenting (and uncommenting) lines 90+91 with a recompilation in between.

**Comment by dennisklein** [comment]
@KlewinS So far I could not reproduce the error by modifying files and recompiling. The only case when I actually trigger the `No dictionary found` error, is, when I add

```cpp
#pragma link C++ class std::vector < o2::TPC::ClusterHardwareContainer8kb > +;
#pragma link C++ class std::vector < o2::TPC::ClusterHardwareContainerFixedSize < 8192 > > +;
```

to the `DataFormatsTPCLinkDef.h` file (or just one of the lines). I also tried without the `+`, but this did not help.

**Comment by dennisklein** [comment]
I tried to verify, if the dictionary is generated or not:

If I add `#pragma link C++ class std::vector < o2::TPC::ClusterHardwareContainer8kb > +;` to `DataFormatsTPCLinkDef.h` file, compile and run:
```
$ alienv enter O2::latest
[O2/latest] ~/alice %> root -l                                                       
root [0] .L sw/fedora27_x86-64/O2/latest/lib/libDataFormatsTPC.so
root [1] auto cl = TClass::GetClass("std::vector<o2::TPC::ClusterHardwareContainer8kb, std::allocator<o2::TPC::ClusterHardwareContainer8kb>>");
root [2] cl->HasDictionary()
(bool) true
root [3] .q
[O2/latest] ~/alice %> root -l
root [0] auto cl = TClass::GetClass("std::vector<o2::TPC::ClusterHardwareContainer8kb, std::allocator<o2::TPC::ClusterHardwareContainer8kb>>");
root [1] cl->HasDictionary()
(bool) false
```

Depending on what `HasDictionary()` [returns](https://github.com/FairRootGroup/FairRoot/blob/dev/base/sink/FairRootFileSink.cxx#L344), you will get the `No dictionary found` error. And if I remove the line from the LinkDef file, I get false in both cases. Works as expected I would say.

Suprisingly, the manual test yields exactly the inverted behaviour of what we see. There must be some additional magic somewhere ... Trying to understand now how this rootmap autoloading facility works.

**Comment by KlewinS** [comment]
So, the lines in DataFormatsTPCLinkDef.h should now stay there or not? I observed the same what you write in the previous comment, therefore I commented them out. But what do you mean with "the manual test"?

**Comment by sawenzel** [comment]
They have to stay (as is the case for any quantity you want to serialize in a ROOT file). 

Your problem is quite subtle and I don't know yet why the runtime doesn't find the dictionary. 

I would be in favour of moving this issue to O2 since it has nothing to do (until proven otherwise) with FairRoot.

**Comment by KlewinS** [comment]
ok, do you want to open an issue in O2 (maybe you have already some additional insights) or shall I do it?

**Comment by dennisklein** [comment]
> I would be in favour of moving this issue to O2 since it has nothing to do (until proven otherwise) with FairRoot.

Agreed. Btw, do you know any documentation for cling/rootcling? I cannot even find a proper list which explains which `#pragma` statements are supported.

> The currently supported way to write a shared_ptr<T> is, like described, to get the raw pointer and put into a static lvalue which is passed to ROOT.

I guess I will open also a separate issue for this aspect, because telling people to use global variables in newly written code worries me a lot. Does it really need to be an in/out parameter in the `RegisterAny` interface?

**Comment by dennisklein** [comment]
> ok, do you want to open an issue in O2 (maybe you have already some additional insights) or shall I do it?

@KlewinS Yes. And please add a reference to this issue (`FairRootGroup/FairRoot#789`). Then please close the issue here. Thx!

**Comment by dennisklein** [comment]
> But what do you mean with "the manual test"?

@KlewinS Sry, I just meant the listing in the same post of mine where I did the check for the presence of the dictionary "manually".

**Comment by KlewinS** [comment]
thanks for the help, issue was moved to O2 https://github.com/AliceO2Group/AliceO2/issues/1186
